### PR TITLE
support local dirs as repos

### PR DESCRIPTION
**Pull request by chaoky**
my use case is that, in react native some packages comes from the `node_modules` dir,
as an alternative, the derivation could simply not fail on missing deps,
Gradle should be able to find these by itself (I think)

### CODE REVIEW

**Comment by CrazyChaoz** [comment]
does react native build with those changes?

i have no clue how react native handles gradle

**Comment by chaoky** [comment]
no, you still need additional configuration, since subprojects repositories cannot be overwritten
https://github.com/status-im/status-mobile/blob/44732e97365dd193a6216640e4c0b3a8e71c2622/nix/pkgs/patch-maven-srcs/default.nix#L19

**Comment by CrazyChaoz** [comment]
are those dependencies not already in a maven repo structure?
if so, why don't you use an additional init script for the final ```gradle build```?
something like
```
gradleInit2 = pkgs.writeText "init.gradle.kts" ''
    projectsLoaded {
        rootProject.allprojects {
            buildscript {
                repositories {
                    maven { url = uri("${./wherever-the-files-are}") }
                }
            }
            repositories {
                maven { url = uri("${./wherever-the-files-are}") }
            }
        }
    }
    settingsEvaluated {
        pluginManagement {
            repositories {
                maven { url = uri("${./wherever-the-files-are}") }
            }
        }
        dependencyResolutionManagement {
            repositoriesMode.set(RepositoriesMode.PREFER_PROJECT)
            repositories {
                maven { url = uri("${./wherever-the-files-are}") }
            }
        }
    }
  '';
  ```
  
 disclaimer: i know you can have multiple init scripts, but idk if they override eachother this way

**Comment by chaoky** [comment]
the derivations fail when deps are not found in `public-maven-repos`, that's the main blocking thing,
but I don't even use them afterwards, since they're already in `node_modules`, and that's where react-native looks by default

gradlePatch.sh
```bash
#!/bin/bash
abs=$(realpath $1)
escp=${abs//\//\\/}
search='\(mavenCentral()\|google()\|jcenter()\|gradlePluginPortal()\)'
replace='maven { url = uri("'"$escp"'") }'
find -L . \
        -name build.gradle -or \
        -name build.gradle.kts -or \
        -name settings.gradle -or \
        -name settings.gradle.kts \
| xargs sed -i "s/$search/$replace/g"
```
and then 
```nix
 '' 
 cp -r ${nodeModules}/node_modules .
 ./gradlePatch.sh ${gradleModules.mvn-repo}
 cd android
 gradle \
        --offline \
        --parallel \
        --console=plain \
        -Dorg.gradle.internal.plugins.portal.url.override=${gradleModules.mvn-repo} \
        -Dorg.gradle.project.android.aapt2FromMavenOverride=$ANDROID_SDK_ROOT/build-tools/34.0.0/aapt2 \
        assembleRelease
 ''
```

**Comment by chaoky** [comment]
I guess I could patch the `verification-metadata` too, but this seems like a more general solution, I'm not a Gradle user tho, so I'm not sure if it's common to requires deps from arbitrary files or if this very specific to react-native

**Comment by CrazyChaoz** [comment]
what is the current gradle build command without nix?

because this project is basiyally a glorified gradle dependency fetcher, but if your deps are already offline there are other methods to include those

**Comment by chaoky** [comment]
that would be only `gradle assembleRelease`, with nix I need
- the `gradlePatch.sh` script to force everything to use `gradle-dot-nix` as a source
-  `org.gradle.internal.plugins.portal.url.override` since some plugins just come from gradle plugin portal and don't care about sources
- `org.gradle.project.android.aapt2FromMavenOverride` is a [kown requirement](https://nixos.wiki/wiki/Android#gradlew)

only some deps comes from `node_modules` not all of them, so this package is very useful as a "glorified gradle dependency fetcher" hahahaha

**Comment by chaoky** [comment]
in the snippet, since I already need `gradlePatch.sh`, I'm also patching my own project with it, instead of using `gradle-init`

**Comment by chaoky** [comment]
for my use case, deps found in `local-repos` don't *have* to be copied into the store, just verifying that those deps are reachable and skipping them would also be fine

**Comment by CrazyChaoz** [comment]
wouldn't modifying your shell script to 
```
replace='maven { url = uri("'"$escp"'") } maven { url = uri("./node_modules") }'
```
work aswell?

**Comment by chaoky** [comment]
I may be misunderstanding you, but the default Gradle config for react native already has something similar, you don't need to tell it to look at `./node_modules`
```gradle
allprojects {
    repositories {
        maven {
            // All of React Native (JS, Obj-C sources, Android binaries) is installed from npm
            url(new File(['node', '--print', "require.resolve('react-native/package.json')"].execute(null, rootDir).text.trim(), '../android'))
        }
        maven {
            // Android JSC is installed from npm
            url(new File(['node', '--print', "require.resolve('jsc-android/package.json', { paths: [require.resolve('react-native/package.json')] })"].execute(null, rootDir).text.trim(), '../dist'))
        }

        google()
        mavenCentral()
        maven { url 'https://www.jitpack.io' }
    }
}
```

**Comment by CrazyChaoz** [comment]
are you sure your issue isn't because the nix sandbox can't resolve / hasn't resolved the ```node_modules``` ?
or did you put them under git?

**Comment by chaoky** [comment]
with this pull request, the derivations works, without the pull request, `gradle-do-nix` fails because some deps could not be downloaded (as they're only available offline, under `node_modules`), 
here's the full example:
```nix
{ pkgs, mkPnpmPackage, gradle-dot-nix }:

let
  nodejs = pkgs.nodejs_21;
  pnpm = pkgs.nodePackages.pnpm;
  java = pkgs.jdk17;
  androidPkgs = pkgs.androidenv.composeAndroidPackages { ... };
  nodeModules = mkPnpmPackage { ... };
  gradleModules = import gradle-dot-nix {
    inherit pkgs;
    gradle-verification-metadata-file = ./android/gradle/verification-metadata.xml;
    local-repos = [ "${nodeModules}/node_modules/jsc-android/dist" ];
  };
  android = pkgs.stdenv.mkDerivation {
    name = "app for android";
    src = ./.;
    nativeBuildInputs = with pkgs; [
      nodejs
      pnpm
      vips
      java
      androidPkgs.androidsdk
      androidPkgs.platform-tools
      (gradle.override { inherit java; })
    ];
    shellHook = ''
      export JAVA_HOME=${java.home}
      export ANDROID_SDK_ROOT=${androidPkgs.androidsdk}/libexec/android-sdk
      export ANDROID_NDK_ROOT=$ANDROID_SDK_ROOT/ndk-bundle/
    '';
    buildPhase = ''
      runHook shellHook
      cp -r ${nodeModules}/node_modules .
      chmod -R +w node_modules
      bash ./gradlePatch.sh ${gradleModules.mvn-repo}
      cd android
      gradle \
        --offline \
        --parallel \
        --console=plain \
        -Dorg.gradle.internal.plugins.portal.url.override=${gradleModules.mvn-repo} \
        -Dorg.gradle.project.android.aapt2FromMavenOverride=$ANDROID_SDK_ROOT/build-tools/34.0.0/aapt2 \
        assembleRelease
    '';
    installPhase = ''
      mkdir -p $out
      cp app/build/outputs/apk/release/* $out
    '';
  };
  web = pkgs.stdenv.mkDerivation { ...  };
in
{
  inherit android web;
  shell = pkgs.mkShell {
    inputsFrom = [ android ];
    packages = with pkgs; [ watchman ];
  };
}
```

**Comment by CrazyChaoz** [comment]
cant you make a seperate derivation that combines the 2 outputs, using a bash script like:

```
#!/bin/bash

mkdir -p "$out"

find "${gradleModules.mvn-repo}" -type f -exec cp --parents {} "$out" \;
find "${nodeModules}/node_modules/jsc-android/dist" -type f -exec cp --parents {} "$out" \;

```

disclaimer: this is ai generated, idk if this is actually working bash code

**Comment by chaoky** [comment]
sorry I don't understand what you're asking, without access to `node_modules` or a option to ignore missing deps, `gradle-dot-nix` will fail to build, so there will be no `gradleModules` to copy from(?), everything else is working fine.
Thanks for making this project, I understand if this pr is considered to be out of scope.
```
error: builder for '/nix/store/zqhaxnczahskwpq6x1ya843r6b5vq458-android-jsc-r250231.aar.drv' failed to produ
ce output path for output 'out' at '/nix/store/zqhaxnczahskwpq6x1ya843r6b5vq458-android-jsc-r250231.aar.drv.
chroot/nix/store/rgqs20bhsm8jk19wq6vhfv3h6wvglq69-android-jsc-r250231.aar'
```

**Comment by CrazyChaoz** [comment]
wait, are the node dependencies in the verification metadata, but not available via ressources?!?

if thats the case i should probably issue warnings, and not fail on missing deps, you are right.

**Comment by chaoky** [comment]
react-native ships a maven repo under `node-modules/jsc-android/dist`, so yeah, since these are just regular Gradle dependencies they end up in the verification file, that's why I said, patching the `verification-metadata.xml` is also an option (if this pr is out of scope e.e) but it would be really good if an option like `local-maven-repos` or  `disk-maven-repos` was available, it feels like a natural extension to `public-maven-repos`, why only fetch using http?

**Comment by CrazyChaoz** [comment]
could you add a usage example to the readme?

File: default.nix [hunk 1]
```diff
@@ -168,6 +185,12 @@ let
   # it changes the project settings (settings.gradle.kts or settings.gradle) to use our repository
   # i'm not sure if we also need repositoriesMode.set(RepositoriesMode.PREFER_PROJECT), but it surely helps
   gradleInit = pkgs.writeText "init.gradle.kts" ''
+    beforeSettings {
```
Comment (chaoky) [inline default.nix:188]: this is not directly related to this issue, it could go in another PR, but is needed for plugins in `includeBuild` to work

File: default.nix [hunk 2]
```diff
@@ -168,6 +185,12 @@ let
...
+        System.setProperty(
+          "org.gradle.internal.plugins.portal.url.override",
+          "${gradle-dependency-maven-repo}"
+        )
+    }
```
Comment (chaoky) [inline default.nix:193]: this is not directly related to this issue, it could go in another PR, but is needed for plugins in `includeBuild` to work

**Comment by CrazyChaoz** [comment]
are you happy with your PR ?
does everything build ?

**Comment by chaoky** [comment]
Besides the unresolved comment, this PR should be ok, everything builds, very minor changes

File: default.nix (hunk 2 above)
Comment (chaoky) [reply default.nix:193]: @CrazyChaoz
